### PR TITLE
Upgrade GraqlSteps (BDD) to Grakn 2.0

### PR DESCRIPTION
**Pull request by alexjpwalker**
## What is the goal of this PR?

Our implementation of BDD, `GraqlSteps`, was out of date, with much of the code being Grakn 1.8 code. We upgraded it to 2.0.

## What are the changes implemented in this PR?

Upgrade GraqlSteps (BDD) to Grakn 2.0

### CODE REVIEW

File: test/behaviour/graql/GraqlSteps.java [hunk 1]
```diff
@@ -22,39 +22,47 @@
...
 import static grakn.client.test.behaviour.util.Util.assertThrowsWithMessage;
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
 
 public class GraqlSteps {
 
-    private static Stream<ConceptMap> answers;
-    private static Number numericAnswer;
+    private static List<ConceptMap> answers;
```
Comment (alexjpwalker) [inline test/behaviour/graql/GraqlSteps.java:62]: `Stream<ConceptMap>` was used for efficiency: allowing us to asynchronously run `match` while executing other steps. However, if we wanted to execute two follow-up steps (e.g: identify the answers in two different ways), that would require re-executing the query.

To mitigate this issue, we now collect the entire `Stream` to a `List` in `get answers of graql match`.

File: test/behaviour/graql/GraqlSteps.java [hunk 2]
```diff
@@ -128,238 +136,226 @@ public void graql_delete_throws_exception(String exception, String deleteQuerySt
...
     public void answer_quantity_assertion(int expectedAnswers) {
-        // TODO
-//        assertEquals(
-//                String.format("Expected [%d] answers, but got [%d]", expectedAnswers, answers.count()),
-//                expectedAnswers, answers.count());
+        assertEquals(String.format("Expected [%d] answers, but got [%d]", expectedAnswers, answers.size()),
+                     expectedAnswers, answers.size());
     }
 
-    @Then("concept identifiers are")
```
Comment (alexjpwalker) [inline test/behaviour/graql/GraqlSteps.java:180]: This step is no longer used. Answer concept matching was simplified to just `uniquely identify answer concepts` a while back.

File: test/behaviour/graql/GraqlSteps.java [hunk 3]
```diff
@@ -128,238 +136,226 @@ public void graql_delete_throws_exception(String exception, String deleteQuerySt
...
-    public void concept_identifiers_are(Map<String, Map<String, String>> identifiers) {
-        for (Map.Entry<String, Map<String, String>> entry : identifiers.entrySet()) {
-            String identifier = entry.getKey();
-            String check = entry.getValue().get("check");
-            String value = entry.getValue().get("value");
+    @Then("uniquely identify answer concepts")
+    public void uniquely_identify_answer_concepts(List<Map<String, String>> answerConcepts) {
```
Comment (alexjpwalker) [inline test/behaviour/graql/GraqlSteps.java:193]: This implementation, and all remaining implementations further down this file, are copied from Grakn Core.

File: test/behaviour/graql/GraqlSteps.java [hunk 4]
```diff
@@ -369,24 +365,44 @@ public AnswerIdentifierGroup(List<Map<String, String>> answerIdentifierTable, Ma
         }
     }
 
-    private boolean matchAnswer(Map<String, String> answerIdentifiers, ConceptMap answer) {
-
-        if (!(answerIdentifiers).keySet().equals(answer.map().keySet())) {
-            return false;
-        }
+//    private boolean matchAnswer(Map<String, String> answerIdentifiers, ConceptMap answer) {
```
Comment (flyingsilverfin) [inline test/behaviour/graql/GraqlSteps.java:368]: what is this comment?

File: test/behaviour/graql/GraqlSteps.java [hunk 5]
```diff
@@ -565,9 +583,38 @@ public boolean check(Concept concept) {
         }
 
         public boolean check(Concept concept) {
-            return concept.isAttribute()
-                    && type.equals(concept.asAttribute().asRemote(tx()).getType().getLabel())
-                    && value.equals(concept.asAttribute().getValue().toString());
+            if (!concept.isAttribute()) {
```
Comment (flyingsilverfin) [inline test/behaviour/graql/GraqlSteps.java:586]: style note: when we can fit into the char limit on one line, we tend to do so and get rid of the {}. However, i there's any block in a series of `if... else if... else` that has to have a multiline {}, all of them use multiline {} to be consistent

**Review by flyingsilverfin — APPROVED** [review]

File: test/behaviour/graql/GraqlSteps.java (hunk 4 above)
Comment (alexjpwalker) [reply test/behaviour/graql/GraqlSteps.java:368]: It's a method used only by Explanation steps, so for consistency with them (& to fix compiler warning), I've commented it out

Comment (flyingsilverfin) [reply test/behaviour/graql/GraqlSteps.java:368]: hmm i can't see the context but is it near the steps that use the method? unclear right now its used for explanations. Might not be clear in core either tbh

Comment (alexjpwalker) [reply test/behaviour/graql/GraqlSteps.java:369]: Added a comment to indicate this

**Pull request merged**